java/cli/src/main/java/org/ray/cli/RayCli.java

```diff
@@ -0,0 +1,326 @@
+
+/**
+ * Ray command line interface
+ */
+public class RayCli {
+
+  private static RayCliArgs rayArgs = new RayCliArgs();
+
+  private static RunManager startRayHead(RayParameters params, PathConfig paths,
+      ConfigReader configReader) {
+    RunManager manager = new RunManager(params, paths, configReader);
+
+    try {
+      manager.startRayHead();
+    } catch (Exception e) {
+      e.printStackTrace();
+      RayLog.core.error("error at RayCli startRayHead", e);
+      throw new RuntimeException("Ray head node start failed, err = " + e.getMessage());
+    }
+
+    RayLog.core.info("Started Ray head node. Redis address: " + manager.info().redisAddress);
+    return manager;
+  }
+
+  private static RunManager startRayNode(RayParameters params, PathConfig paths,
+      ConfigReader configReader) {
+    RunManager manager = new RunManager(params, paths, configReader);
+
+    try {
+      manager.startRayNode();
+    } catch (Exception e) {
+      e.printStackTrace();
+      RayLog.core.error("error at RayCli startRayNode", e);
+      throw new RuntimeException("Ray work node start failed, err = " + e.getMessage());
+    }
+
+    RayLog.core.info("Started Ray work node.");
+    return manager;
+  }
+
+  private static RunManager startProcess(CommandStart cmdStart, ConfigReader config) {
+    PathConfig paths = new PathConfig(config);
+    RayParameters params = new RayParameters(config);
+
+    // Get the node IP address if one is not provided.
+    //if (params.node_ip_address.length() == 0) {
+    //    params.node_ip_address = NetworkUtil.getIpAddress(null);
+    //}
+    RayLog.core.info("Using IP address " + params.node_ip_address + " for this node.");
+    RunManager manager;
+    if (cmdStart.head) {
+      manager = startRayHead(params, paths, config);
+    } else if (cmdStart.work) {
+      manager = startRayNode(params, paths, config);
+    } else {
+      throw new RuntimeException(
+          "Ray must be started with at least a role. Seeing --help for more details.");
+    }
+    return manager;
+  }
+
+  private static void start(CommandStart cmdStart, ConfigReader reader) {
+    RayParameters params = new RayParameters(reader);
+
+    RunManager manager = null;
+
+    manager = startProcess(cmdStart, reader);
+
+    // monitoring all processes, throwing an exception when any process fails
+    while (true) {
+      try {
+        Thread.sleep(5000);
+      } catch (InterruptedException e) {
+      }
+
+      HashSet<RunInfo.ProcessType> excludeTypes = new HashSet<>();
+      if (!manager.checkAlive(excludeTypes)) {
+
+        //ray components fail-over
+        RayLog.core.error("Something error in Ray processes.");
+
+        if (!manager.tryRecoverDeadProcess()) {
+          RayLog.core.error("Restart all the processes in this node.");
+          manager.cleanup(true);
+          manager = startProcess(cmdStart, reader);
+        }
+      }
+    }
+  }
+
+  private static void stop(CommandStop cmdStop) {
```

Comment:
`killall` and `ps/grep/kill` could be dangerous`. For example, if I have a `vim global_scheduler.cc`, the vim process will also be killed.

Since the RunManager already have records of all process ids, we can kill the processes by the ids.

Reply:
Good suggestion. Currently cli doesn't have this information, we can do this in a later phase by implementing a service which has all this information, and let cli to talk to this service.